src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs

```diff
@@ -114,14 +120,44 @@ private BoundStatement AppendJumpToCurrentFinallyOrExit(BoundStatement node)
                 GenerateJumpToCurrentFinallyOrExit());
         }
 
+        protected override BoundBinaryOperator ShouldEnterFinallyBlock()
+        {
+            // We should skip the finally block when:
+            // - the state is 0 or greater (we're suspending on an `await`)
+            // - the state is -3, -4 or lower (we're suspending on a `yield return`)
+            // We don't care about state = -2 (method already completed)
+
+            // So we only want to enter the finally when the state is -1
+            return F.IntEqual(F.Local(cachedState), F.Literal(StateMachineStates.NotStartedStateMachine));
+        }
+
         #region Visitors
 
+        /// <summary>
+        /// Lower the body, adding an entry state (-3) at the start,
+        /// so that we can differentiate a async-iterator that was never moved forward with MoveNextAsync()
```

Comment:
>a [](start = 41, length = 1)

`an`?

Reply:
I'll include this fix in my next async-streams PR.

---
In reply to: [245378700](https://github.com/dotnet/roslyn/pull/31764#discussion_r245378700) [](ancestors = 245378700)